control/controlclient/direct.go

```diff
@@ -1303,3 +1303,52 @@ func (c *Direct) SetDNS(ctx context.Context, req *tailcfg.SetDNSRequest) error {
 
 	return nil
 }
+
+// tsmpPing sends a Ping to pr.IP, and sends an http request back to pr.URL
+// with ping response data.
+func tsmpPing(logf logger.Logf, c *http.Client, pr *tailcfg.PingRequest, pinger Pinger) {
+	if pr.URL == "" {
+		logf("invalid PingRequest with no URL")
+		return
+	}
+	if pr.IP.IsZero() {
+		logf("invalid PingRequest with no proper IP got %v", pr.IP)
+		return
+	}
+
+	now := time.Now()
+	pinger.Ping(pr.IP, true, func(res *ipnstate.PingResult) {
+		if res.Err != "" {
+			logf(res.Err)
+			return
+		}
+		duration := time.Since(now)
+		logf("TSMP ping completed in %v seconds. pinger.Ping took %v", res.LatencySeconds, duration)
+		ctx, cancel := context.WithTimeout(context.Background(), 15*time.Second)
+		defer cancel()
+
+		jsonPingRes, err := json.Marshal(res)
+		if err != nil {
+			logf("%v", err)
+			return
+		}
+		// Send the results of the Ping, back to control URL.
+		req, err := http.NewRequestWithContext(ctx, "POST", pr.URL, bytes.NewBuffer(jsonPingRes))
+		if err != nil {
+			logf("http.NewRequestWithContext(%q): %v", pr.URL, err)
+			return
+		}
+		if pr.Log {
+			logf("tsmpPing: sending ping to %v ...", pr.URL)
```

Comment:
should that be pr.IP, or should the log be "sending ping results to %v"?

Reply:
changed to ping results to %v